formula/src/main/java/com/instacart/formula/FormulaRuntime.kt

```diff
@@ -69,47 +69,60 @@ class FormulaRuntime<Input : Any, Output : Any>(
             effectQueue.addLast(it)
         }
 
-        run(shouldEvaluate = evaluate)
+        run(evaluate = evaluate)
+    }
+
+    override fun onPostponedTransition(event: Event) {
+        event.dispatch()
     }
 
-    private fun forceRun() = run(shouldEvaluate = true)
+    private fun forceRun() = run(evaluate = true)
 
     /**
      * Performs the evaluation and execution phases.
      *
-     * @param shouldEvaluate Determines if evaluation needs to be run.
+     * @param evaluate Determines if evaluation needs to be run.
      */
-    private fun run(shouldEvaluate: Boolean) {
+    private fun run(evaluate: Boolean) {
+        if (isEvaluating) return
+
         try {
-            val freshRun = !isExecuting
-            if (freshRun) {
-                inspector?.onRunStarted(shouldEvaluate)
-            }
+            runFormula(evaluate)
+            emitOutputIfNeeded(isInitialRun = false)
+        } catch (e: Throwable) {
+            isEvaluating = false
+
+            manager?.markAsTerminated()
+            onError(e)
+            manager?.performTerminationSideEffects()
+        }
+    }
 
-            val manager = checkNotNull(manager)
-            val currentInput = checkNotNull(input)
+    private fun runFormula(evaluate: Boolean) {
+        val freshRun = !isExecuting
+        if (freshRun) {
+            inspector?.onRunStarted(evaluate)
+        }
 
-            if (shouldEvaluate && !manager.terminated) {
-                evaluationPhase(manager, currentInput)
-            }
+        val manager = checkNotNull(manager)
+        val currentInput = checkNotNull(input)
+
+        if (evaluate && !manager.terminated) {
+            isEvaluating = true
+            evaluationPhase(manager, currentInput)
+            isEvaluating = false
+        }
 
+        if (evaluate || effectQueue.isNotEmpty()) {
```

Comment:
`executionRequested` does not seem to be used anymore

Reply:
Removed